Table pprint() to handle unicode in Python 2

pllim opened the pull request:
This fixes #4944. That is, `pprint()` now handles unicode in Python 2 properly. Also expanded test to include special character and rearranged Python 2/3 logic to handle Python 4.

p.s. I optimistically milestoned this to 1.2. Feel free to remove milestone if it is not possible.

p.p.s. My Emacs also automatically removed trailing whitespace.

Example use case:
```python
In [1]: from astropy.table import Table, Column

In [2]: tab = Table()

In [3]: tab.add_column(Column(data=['blah','bläh'], name='ok'))

In [4]: tab.pprint()
 ok 
----
blah
bläh
```

saimn commented:
This remind me about a similar bug with a FITS table (read with `astropy.table.Table`) and I forgot to investigate ... Testing with this table and your PR fix the issue, so :+1:

saimn commented on astropy/table/pprint.py:
```diff
@@ -13,19 +13,26 @@
         if isinstance(val, bytes):
             return val.decode('utf-8')
         else:
             return str(val)
```
The block could be simplified with a `return text_type(val) if six.PY2 else str(val)`.

pllim replied:
I thought about that, but doesn't that make it less efficient, since it has to compare at every call? Doing it this way, the function is set once it determines whether user is running Python 2 or 3.

saimn replied:
Ah yes, true !

taldcroft replied:
In Py3 and beyond (presumably), `six.text_type` is `str`, so can you remove the `if six.PY2` block completely?

saimn commented on astropy/table/pprint.py:
```diff
@@ -13,19 +13,26 @@
 from ..utils.data_info import dtype_info_name
 
-if six.PY3:
+__all__ = []
```
I *think* having an empty `__all__` is like not having it ?

pllim replied:
Is it? In my past experience, not having it equals getting a bunch of crap when doing "from module import *". But this has nothing to do with the fix, so I'm okay with removing it if desired.

saimn replied:
Or maybe add `'TableFormatter'` if you prefer ?

pllim replied:
I'd rather not because of this in `docs/table/index.rst`:

```
.. automodapi:: astropy.table
```

Adding something in `__all__` might change what is listed in the API doc, which is not the intend of this PR. So, either I remove it or keep it as empty list.

saimn replied:
It should be ok for `table.pprint` as it seems not imported/documented, but yeah, better safe than sorry. I let you judge on the final choice, it was a detail.

taldcroft commented on astropy/table/tests/test_pprint.py:
```diff
@@ -528,13 +529,24 @@ def test_pprint_npfloat32():
     """
-    val = bytes('val', encoding='utf-8') if PY3 else 'val'
-    dat = np.array([(val,)], dtype=[(str('col'), 'S3')])
+    val = 'val' if PY2 else bytes('val', encoding='utf-8')
```
Can you get rid of one or both of the xfails below by setting `val = str('val')`?  In this case it really should work always, no?

taldcroft commented:
Thanks @patti and @saimn.  This looks basically good with some minor comments/questions.

pllim commented:
@taldcroft ans @saimn , I think I have addressed all the comments. Thanks for the tips, @taldcroft , indeed casting `str` removed the necessity to mark `xfail` in Python 2.

taldcroft commented:
@pllim - can you try applying this to earlier versions (1.0.x, 1.1.x) to see if it merges cleanly and passes tests?   I just re-milestoned to 1.3 but realized since this is a bug fix it should be backported as far as possible.

@eteq @astrofrog - what's the status of non-critical bug fixes for 1.2.0?  Still being accepted or no?  This can wait until 1.2.1 but there is no such milestone visible yet.

pllim commented:
@taldcroft , sure. But looks like GitHub is partially down right now. I'll do it when GitHub lets me.

pllim commented:
@taldcroft , do you want me to open the backporting to v1.0.x and v1.1.x as separate PRs? I already have branches for them, so it is just a matter of clicking "submit PR" button. If not, they are at https://github.com/pllim/astropy/tree/issue4946-v1.0.x and https://github.com/pllim/astropy/tree/issue4946-v1.1.x, respectively.

taldcroft commented:
@pllim - our illustrious package manager(s) generally take care of actually backporting outside of github, but it is good to know in advance that the changes can be merged with no conflicts and that tests pass.

pllim commented:
Tests seem okay.

**v1.0.x**

Travis: https://travis-ci.org/pllim/astropy/builds/132333404
Appveyor: https://ci.appveyor.com/project/pllim/astropy/build/1.0.25

**v1.1.x**

Travis: https://travis-ci.org/pllim/astropy/builds/132344558
Appveyor: https://ci.appveyor.com/project/pllim/astropy/build/1.0.24

taldcroft commented:
@pllim - awesome!  Did you have to do anything special to set up CI on your own branches?  Or just following the GitHub instructions?

pllim commented:
For Travis, I think I added it via my fork project settings. For Appveyor, I think I had to log into my own Appveyor account (not Astropy's) and add my own fork project there. Then, have to make sure they are both marked as Active in GitHub fork project settings page. Then a new commit to my fork should kick them off (or manually test payload deployment). Hope this helps.

bsipocz commented:
@pllim - Thanks, this is very useful information. I use travis all the time on my forks, but never managed to do it on appveyor (but also never bothered enough to go after a solution).

pllim commented:
Now you know, and knowledge is power. :wink:

taldcroft commented:
OK, I changed the milestone to 1.0.10.
